sextant/src/npk.rs

```diff
@@ -512,30 +523,30 @@ fn write_verity_header(
     assert_eq!(fsimg_size % BLOCK_SIZE as u64, 0);
     let data_blocks = fsimg_size / BLOCK_SIZE as u64;
 
-    /* ['verity', 1, 1, uuid.gsub('-', ''), 'sha256', 4096, 4096, data_blocks, 32, salt, '']
-     * .pack('a8 L L H32 a32 L L Q S x6 a256 a3752')
-     * (https://gitlab.com/cryptsetup/cryptsetup/-/wikis/DMVerity#verity-superblock-format)
-     * (https://ruby-doc.org/core-2.7.1/Array.html#method-i-pack) */
     let mut fsimg = OpenOptions::new()
         .write(true)
         .append(true)
         .open(&fsimg_path)
         .with_context(|| format!("Cannot open '{}'", &fsimg_path.display()))?;
-    fsimg.write_all(b"verity")?;
-    fsimg.write_all(&[0_u8, 0_u8])?;
-    fsimg.write_all(&1_u32.to_ne_bytes())?;
-    fsimg.write_all(&1_u32.to_ne_bytes())?;
+
+    // https://gitlab.com/cryptsetup/cryptsetup/-/wikis/DMVerity#verity-superblock-format
+    const VERITY_SIGNATURE: &[u8; 8] = b"verity\x00\x00";
+    const HASH_ALG_NAME: &[u8; 6] = b"sha256";
+    fsimg.write_all(VERITY_SIGNATURE)?;
+    fsimg.write_all(&1_u32.to_ne_bytes())?; // superblock version
+    fsimg.write_all(&1_u32.to_ne_bytes())?; // hash type 'normal'
     fsimg.write_all(&hex::decode(uuid.to_string().replace("-", ""))?)?;
-    fsimg.write_all(b"sha256")?;
+    fsimg.write_all(HASH_ALG_NAME)?;
     fsimg.write_all(&[0_u8; 26])?;
-    fsimg.write_all(&4096_u32.to_ne_bytes())?;
-    fsimg.write_all(&4096_u32.to_ne_bytes())?;
-    fsimg.write_all(&data_blocks.to_ne_bytes())?;
-    fsimg.write_all(&32_u16.to_ne_bytes())?;
-    fsimg.write_all(&[0_u8; 6])?;
+    fsimg.write_all(&4096_u32.to_ne_bytes())?; // data block in bytes
+    fsimg.write_all(&4096_u32.to_ne_bytes())?; // hash block in bytes
+    fsimg.write_all(&data_blocks.to_ne_bytes())?; // number of data blocks
+    fsimg.write_all(&32_u16.to_ne_bytes())?; // salt size (SHA256 output length)
+    fsimg.write_all(&[0_u8; 6])?; // padding
     fsimg.write_all(&salt.to_vec())?;
-    fsimg.write_all(&vec![0_u8; 256 - salt.len()])?;
-    fsimg.write_all(&[0_u8; 3752])?;
+    fsimg.write_all(&vec![0_u8; 256 - salt.len()])?; // padding
+    const BYTES_WRITTEN: usize = 8 + 4 + 4 + 16 + 6 + 26 + 4 + 4 + 8 + 2 + 6 + 256;
```

Comment:
maybe we can use `write_vectored` here? (https://doc.rust-lang.org/std/io/trait.Write.html#method.write_vectored)
![image](https://user-images.githubusercontent.com/99541/98075773-469e9a80-1e6d-11eb-917b-d762a694d6c4.png)
this could give us the bytes written directly

Reply:
I ended up using a vector as a superblock buffer and extending it by each field individually. That way, we have access to the number of bytes to be written. Using IoSlice did not work as we are writing a lot of temporaries.